src/keys-builder/template/utils.ts

```diff
@@ -98,3 +109,72 @@ export function isSupportedNode<Predicates extends any[]>(
 ): node is GuardedType<Predicates[number]> {
   return predicates.some((predicate) => predicate(node));
 }
+
+export function isBlockWithChildren(
+  node: unknown,
+): node is { children: TmplAstNode[] } {
+  return (
+    node instanceof TmplAstDeferredBlockError ||
+    node instanceof TmplAstDeferredBlockLoading ||
+    node instanceof TmplAstDeferredBlockPlaceholder ||
+    node instanceof TmplAstForLoopBlockEmpty ||
+    node instanceof TmplAstIfBlockBranch ||
+    node instanceof TmplAstSwitchBlockCase
+  );
+}
+
+export function isTmplAstForLoopBlock(
+  node: unknown,
+): node is TmplAstForLoopBlock {
+  return node instanceof TmplAstForLoopBlock;
+}
+
+export function isTmplAstDeferredBlock(
+  node: unknown,
+): node is TmplAstDeferredBlock {
+  return node instanceof TmplAstDeferredBlock;
+}
+
+export function isTmplAstIfBlock(node: unknown): node is TmplAstIfBlock {
+  return node instanceof TmplAstIfBlock;
+}
+
+export function isTmplAstSwitchBlock(
+  node: unknown,
+): node is TmplAstSwitchBlock {
+  return node instanceof TmplAstSwitchBlock;
+}
+
+function isNotNull<T>(input: T | null): input is T {
+  return input !== null;
+}
+
+/**
+ * Returns children nodes of the given node if it's a block node.
+ */
+export function getChildrendNodesIfBlock(node: TmplAstNode): TmplAstNode[] {
```

Comment:
Please separate this into `isBlockNode` and `resolveBlockChildren`
Once separated we can make `node.children` the default return value of `resolveBlockChildren` and have something like this:
```
    if (isBlockNode(node)) {
      traverse(resolveBlockChildNodes(node), containers, config);
      continue;
    }
```

Reply:
This means we'll have to run the if tests twice. Not very optimizing, is it?

What's more, if I return `node.children` without doing an if, I get a typing error.

```ts
export function isBlockNode(node: TmplAstNode): boolean {
  // test here for the first time
  return isTmplAstIfBlock(node) ||
    isTmplAstForLoopBlock(node) ||
    isTmplAstDeferredBlock(node) ||
    isTmplAstSwitchBlock(node) ||
    isBlockWithChildren(node);
}

export function resolveBlockChildNodes(node: TmplAstNode): TmplAstNode[] {
  // test again
  if (isTmplAstIfBlock(node)) {
    return node.branches;
  }

  if (isTmplAstForLoopBlock(node)) {
    return node.empty ? [...node.children, node.empty] : node.children;
  }

  if (isTmplAstDeferredBlock(node)) {
    // Typing errors here
    return [
      ...node.children,
      ...[node.loading, node.error, node.placeholder].filter(Boolean),
    ];
  }

  if (isTmplAstSwitchBlock(node)) {
    return node.cases;
  }

   // Typing errors here
  return node.children;
}

```

Reply:
1. We don't have any performance considerations here and this isn't some very heavy calculation that I would agree we need to optimize, I'd take readability.
2. What's the type error?

Reply:
Done